examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+                                                    bool assemble_rhs)
+  {
+    if (assemble_matrix)
+      {
+        system_matrix    = 0;
+      }
+
+    if (assemble_rhs)
+      {
+        system_rhs       = 0;
+      }
+
+    QGauss<dim>   quadrature_formula(degree+2);
+
+    FEValues<dim> fe_values (fe,
+                             quadrature_formula,
+                             update_values |
+                             update_quadrature_points |
+                             update_JxW_values |
+                             update_gradients );
+
+    const unsigned int   dofs_per_cell = fe.dofs_per_cell;
+    const unsigned int   n_q_points    = quadrature_formula.size();
+
+    const FEValuesExtractors::Vector velocities (0);
+    const FEValuesExtractors::Scalar pressure (dim);
+
+    FullMatrix<double>   local_matrix (dofs_per_cell, dofs_per_cell);
+    Vector<double>       local_rhs    (dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+    // For the linearized system, we create temporary storages for present velocity
+    // and gradient, and present pressure. In practice, they are
+    // all obtained through their shape functions at quadrature points.
+
+    std::vector<Tensor<1, dim> >  present_velocity_values    (n_q_points);
+    std::vector<Tensor<2, dim> >  present_velocity_gradients (n_q_points);
+    std::vector<double>           present_pressure_values    (n_q_points);
+
+    std::vector<double>           div_phi_u                 (dofs_per_cell);
+    std::vector<Tensor<1, dim> >  phi_u                     (dofs_per_cell);
+    std::vector<Tensor<2, dim> >  grad_phi_u                (dofs_per_cell);
+    std::vector<double>           phi_p                     (dofs_per_cell);
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+
+    for (; cell!=endc; ++cell)
+      {
+        fe_values.reinit(cell);
+
+        local_matrix = 0;
+        local_rhs    = 0;
+
+        fe_values[velocities].get_function_values(evaluation_point,
+                                                  present_velocity_values);
+
+        fe_values[velocities].get_function_gradients(evaluation_point,
+                                                     present_velocity_gradients);
+
+        fe_values[pressure].get_function_values(evaluation_point,
+                                                present_pressure_values);
+
+        // The assembly is similar to step-22. An additional term with gamma as the coefficient
+        // is the Lagrangian Argument(AL), which is assembled via grad-div stabilization.
+        // As we discussed in the introduction, the 1,1-block of system matrix should be
+        // zero. Since the pressure mass matrix is used while creating the preconditioner,
+        // we assemble it here and then move it into a separate SparseMatrix at the end (same as in step-22).
```

Comment:
Nothing to do here, but I'm going to note that in hindsight I consider it a mistake to put the pressure mass matrix into this block. That's because you can't just apply a gmres or direct solver to the matrix any more, given that it no longer contains the linear system. 

It is also not true that the block is zero: it contains entries for hanging nodes.